File: matlab/src/error_prediction.m
function [AMPL,MSE,PSNR] = error_prediction(T,m,n,M,N,varargin)
%%
% Usage: [AMPL,MSE,PSNR] = error_prediction(T,m,n,M,N,Name,Value)
% 
% Input(s)/Output(s):
%
%   T      : (matrix of double) translation vector, must have exactly two
%            columns, i.e., T = [tx,ty], where tx and ty denote the
%            horizontal and vertical components of the translation vectors
%   m      : (scalar positive) width of the low-resolution domain 
%   n      : (scalar positive) height of the low-resolution domain
%   M      : (scalar >= m) width of the high-resolution domain
%   N      : (scalar >= n) height of the high-resolution domain
% 
%   AMPL   : (matrix of double) error amplification coefficient (Fourier
%            domain)
%   MSE    : (scalar, this output is only available when the optional input
%            'sigma' is provided) predicted mean-square-error
%   PSNR   : (scalar, this output is only available when the optional input
%            'sigma' is provided) peak signal-to-noise-ratio computed as
%            PSNR = 10*log10(peakval^2/MSE)
% 
% Optional Name-Value pair arguments:
%
%   ['sigma',sigma]     : (scalar double) set the level of noise (standard
%                         deviation) corrupting the low-resolution sequence
%
%   ['peakval',peakval] : (scalar, default peakval = 255.) peak signal
%                         value used for calculating the predicted peak
%                         signal-to-noise ratio from the predicted
%                         mean-square-error 
%   ['eps',e]           : (scalar double, default e = 1e-9) threshold for
%                         the singular values in the pseudo-inversion
%                         routine (treat as zero all singular values less 
%                         than e when computing the pseudo inverse of a
%                         matrix)
%   ['noshift,nflag]    : (scalar logical, default nflag = false) by
%                         default (nflag = false) the (0,0) frequency
%                         corresponds to the center pixel of AMPL, set 
%                         nflag = true to keep the (0,0) frequency on the
%                         top-left corner of the image (FFTW convention)
% 
% Description: prediction of the super-resolution reconstruction quality
% 

%% Control number of inputs
if(nargin < 5)
    help error_prediction;
    error('Incorrect number of input(s)');
end

%% Parser (manual implementation because MATLAB parser is not case sensitive (inputs (m,M), (n,N) not allowed)

% set default values for optional arguments
sig = []; 
AMPL = []; 
peakval = 255; 
e = 1e-9; 
nflag = false; 

% retrieve optional arguments
for k = 0:floor(numel(varargin)/2)-1
    switch lower(varargin{2*k+1})
        case 'sigma'
            sig = varargin{2*k+2};
        case 'peakval'
            peakval = varargin{2*k+2}; 
        case 'eps'
            e = varargin{2*k+2}; 
        case 'noshift'
            nflag = varargin{2*k+2}; 
        otherwise
            if(isstring(varargin{2*k+1}) || ischar(varargin{2*k+1}))
                error("'%s' is not a recognized parameter. For a list of valid name-value pair arguments, see the documentation for this function.",varargin{2*k+1}); 
            else
                error("Expected a string scalar or character vector for the parameter name."); 
            end
    end
end

%% consistency checks

% check number of elements in varargin (Name-value pair arguments)
if(bitand(numel(varargin),1))
    if(isstring(varargin{end}) || ~ischar(varargin{end}))
        error("No value was given for '%s'. Name-value pair arguments require a name followed by a value.",varargin{end});
    else
        error("Expected a string scalar or character vector for the parameter name."); 
    end
end

% input T (matrix of two double real numbers)
if(~isreal(T) || size(T,2) ~= 2)
    help error_prediction;
    error('input ''T'' must have exactly two columns of double real numbers');
end

% input m (scalar positive, no decimal part)
if(~isreal(m) || ~isscalar(m) || m ~= floor(m))
    help error_prediction;
    error('input m must be a real positive scalar number without decimal part (m == floor(m), m > 0)');
end

% input n (scalar positive, no decimal part)
if(~isreal(n) || ~isscalar(n) || n ~= floor(n))
    help error_prediction;
    error('input n must be a real positive scalar number without decimal part (n == floor(n), n > 0)');
end

% input M (scalar >= m)
if(~isreal(M) || ~isscalar(M) || M ~= floor(M) || m > M)
    help error_prediction;
    error('input M must be a real scalar number, without decimal part (M == floor(M)), grater than or equal to m (M >= m)');
end

% input N (scalar, no decimal part, >= n)
if(~isreal(N) || ~isscalar(N) || N ~= floor(N) || n > N)
    help error_prediction;
    error('input N must be a real scalar number, without decimal part (N == floor(N)), greater than or equal to n (N >= n)');
end

% optional input A (matrix of double, size(A) == [N,M])
if(~isempty(AMPL) && (~isreal(AMPL) || size(AMPL,1) ~= N || size(AMPL,2) ~= M))
    help error_prediction;
    error('optional input ''AMPL'' must be a real matrix of double such as size(AMPL) == [N,M]');
end

% optional input peakval (scalar double)
if(~isscalar(peakval) || ~isreal(peakval))
    help error_prediction;
    error('optional input ''peakval'' must be a scalar double');
end

% optional input sigma (scalar double >= 0)
if(~isempty(sig) && (~isscalar(sig) || ~isreal(sig) || sig < 0))
    help error_prediction;
    error('optional input ''sigma'' must be a nonnegative scalar double');
end

% optional output MSE & PSNR (need to check that optional input 'sigma' is provided)
if((nargout >= 2) && isempty(sig))
    help error_prediction;
    error('optional input ''sigma'' is mandatory to compute the MSE and PSNR outputs');
end

% optional input 'eps'
if(~isreal(e) || ~isscalar(e))
    help error_prediction;
    error('optional input ''eps'' must be a scalar real number');
end

% optional input 'noshift'
if(~islogical(nflag) || ~isscalar(nflag))
    help error_prediction;
    error('optional input ''noshift'' must be a scalar logical');
end

%% CORE OF THE MODULE

% reduce as much as possible domain dimensions (M,m) and (N,n) while
% keeping constant the super-resolution factors zx = M/m and zy = N/n
gx = gcd(M,m); 
gy = gcd(N,n); 
M = M/gx; m = m/gx; % reduce M and m by a factor gx (does not change the ratio zx = M/m)
N = N/gy; n = n/gy; % reduce N and n by a factor gy (does not change the ratio zy = N/n)

% if not provided, compute the error amplification coefficients
if(isempty(AMPL))
    
    AMPL = zeros(N,M);
    
    % compute super-resolution factors 
    zx = M/m; 
    zy = N/n;
    
    % compute the frequency coordinate grid associated to the low-resolution domain
    a = ifftshift(-floor(m/2)+(0:m-1));
    b = ifftshift(-floor(n/2)+(0:n-1));
    [A,B] = meshgrid(a,b);
    A = A(:); B = B(:);
    
    % compute the frequency coordinate grid associated to the high-resolution domain
    alf = ifftshift(-floor(M/2)+(0:M-1));
    bet = ifftshift(-floor(N/2)+(0:N-1));
    [ALF,BET] = meshgrid(alf,bet);
    ALF = ALF(:); BET = BET(:);
    
    % compute ramp-phasis terms
    L = size(T,1); 
    phi = @(alf,bet,dx,dy) exp(-2*1i*pi*(alf*dx/M + bet*dy/N));
    F = reshape(phi(ALF,BET,zx*T(:,1)',zy*T(:,2)'),[N,M,L]);
    
    % compute bounds for the sets P(A) and Q(B)
    PMIN = ceil(-(M+2*A)/(2*m));
    PMAX = ceil((M-2*A)/(2*m))-1;
    QMIN = ceil(-(N+2*B)/(2*n));
    QMAX = ceil((N-2*B)/(2*n))-1;
    
    % precompute block matrices
    M1_PINV = pinv(compute_blockmatrix(T,1/(zx*zy),floor(zx),floor(zy)),e);
    if(zx ~= floor(zx))
        M2_PINV = pinv(compute_blockmatrix(T,1/(zx*zy),ceil(zx),floor(zy)),e);
    end
    if(zy ~= floor(zy))
        M3_PINV = pinv(compute_blockmatrix(T,1/(zx*zy),floor(zx),ceil(zy)),e);
    end
    if(zx ~= floor(zx) && zy ~= floor(zy))
        M4_PINV = pinv(compute_blockmatrix(T,1/(zx*zy),ceil(zx),ceil(zy)),e);
    end
    
    % main loop
    for idfreq = 1:m*n
        
        % extract frequency (a,b) and its corresponding [pmin(a),pmax(a),qmin(b),qmax(b)]
        a = A(idfreq);
        b = B(idfreq);
        pmin = PMIN(idfreq);
        pmax = PMAX(idfreq);
        qmin = QMIN(idfreq);
        qmax = QMAX(idfreq);
        size_p = pmax - pmin + 1;
        size_q = qmax - qmin + 1;
        
        % store in lexicographical order the elements (p,q) of P(a) x Q(b)
        [P,Q] = meshgrid(pmin:pmax,qmin:qmax);
        P = P(:); Q = Q(:);
        
        % retrieve all frequencies (alf,bet) that are aliased at position (a,b)
        % in the low-resolution frequency domain
        alf = a+P*m;
        bet = b+Q*n;
        
        % compute amplification coefficients
        if(floor(zx) == size_p && floor(zy) == size_q)
            AMPL(mod(M+alf,M)*N+mod(N+bet,N)+1) = sqrt(sum(abs(M1_PINV*F(1 + ((0:L-1)*M*N + mod(M+alf,M)*N + mod(N+bet,N)))).^2,2)/(zx*zy));
        elseif(floor(zx) + 1 == size_p && floor(zy) == size_q)
            AMPL(mod(M+alf,M)*N+mod(N+bet,N)+1) = sqrt(sum(abs(M2_PINV*F(1 + ((0:L-1)*M*N + mod(M+alf,M)*N + mod(N+bet,N)))).^2,2)/(zx*zy));
        elseif(floor(zx) == size_p && floor(zy) + 1 == size_q)
            AMPL(mod(M+alf,M)*N+mod(N+bet,N)+1) = sqrt(sum(abs(M3_PINV*F(1 + ((0:L-1)*M*N + mod(M+alf,M)*N + mod(N+bet,N)))).^2,2)/(zx*zy));
        elseif(floor(zx) + 1 == size_p && floor(zy) + 1 == size_q)
            AMPL(mod(M+alf,M)*N+mod(N+bet,N)+1) = sqrt(sum(abs(M4_PINV*F(1 + ((0:L-1)*M*N + mod(M+alf,M)*N + mod(N+bet,N)))).^2,2)/(zx*zy));
        end
        
    end
end

% compute the predicted MSE and PSNR
if(nargout >= 2)
    MSE = sig^2 * sum(AMPL(:).^2) / (M*N);
end
if(nargout >= 3)
    PSNR = 10*log10(peakval^2/MSE);
end

% perform nearest neighbor zooming with factor (gx,gy) to get the
% amplification map at the initial size
AMPL = kron(fftshift(AMPL),ones(gy,gx)); 
if(nflag) 
    AMPL = ifftshift(AMPL); 
end

end

File: c/src/error-prediction.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

#define ASSERT_ALLOC(cmd) if(NULL == (cmd)) { printf("Not enough memory\n"); exit(EXIT_FAILURE); }

/* external modules */
extern int save_monopage_striped_tiff_from_double(char*,double*,int,int,char,int,char*); // see source file 'tiffwrite.c'
extern int getasciitranslations(double**,double**,int*,char*,char); // see source file 'ascii.c'
extern int error_prediction(double*,double*,double*,double*,double*,double*,int,int,int,int,int,double,double,char,char); // see source file 'error-prediction_kernel.c'

/* usage displayer */
static void display_usage()
{
  printf("\nPrediction of the super-resolution reconstruction quality.\n\n");
  printf("Usage: error-prediction [-ftype type] [-M M] [-zx zx] [-N N] [-zy zy] [-A A] [-s sigma] [-e eps] [-noshift] [-v] m n T\n\n");
  printf(" -ftype type   : (default float) datatype of the output TIFF images, possible\n");
  printf("                 choices are {uint8,int8,uint16,int16,uint32,int32,float}\n");
  printf(" -M M          : (default 2 * width of in) set width of the ouptut image...\n");
  printf(" -zx zx        : ... or set X-axis super-resolution factor (double >= 1.)\n");
  printf(" -N N          : (default 2 * height of in) set height of the output image...\n");
  printf(" -zy zy        : ... or set Y-axis super-resolution factor (double >= 1.)\n");
  printf(" -A A          : output error amplification map (Fourier domain)\n");
  printf(" -s sigma      : noise level (standard deviation) of the sequence to process,\n");
  printf("                 this input is mandatory to display the predicted MSE and PSNR\n");
  printf(" -p peakval    : (default 255.) set the peak-value for the PSNR\n");
  printf(" -e eps        : (default 1e-9) threshold for the singular values in the\n");
  printf("                 pseudo-inversion routine\n");
  printf(" -noshift      : do not shift output A of half the image size (put the (0,0)\n");
  printf("                 frequency at the top-left corner of A instead of its center)\n");
  printf(" -v            : enable verbose mode\n"); 
  printf(" m             : width of the low-resolution image domain\n"); 
  printf(" n             : height of the low-resolution image domain\n"); 
  printf(" T             : input sequence of displacements (ASCII format, two-columns)\n");
  printf(" screen output : (with option -s only) predicted peak signal-to-noise ratio (PSNR)\n");
  printf(" screen output : (with option -s only) predicted mean square error (MSE)\n\n");
}

/* command line interface */
int main(int argc, char **argv)
{
  char *datatype=NULL,*fname_in=NULL,*fname_T=NULL,*fname_A=NULL,*M_value=NULL,*N_value=NULL,*zx_value=NULL,*zy_value=NULL,*m_value=NULL,*n_value=NULL,*sigma_value=NULL,*peak_value=NULL,*eps_value=NULL,vflag=0,noshiftflag=0;
  double *dx=NULL,*dy=NULL,*A=NULL,*Asmall=NULL,zx,zy,zx_tmp,zy_tmp,sigma,mse,psnr,peakval,eps;
  int err,k,type,nx,ny,ntranslations,Nx,Ny,gx,gy,u,v,r,x,y,adr,Nxsmall,Nysmall,nxsmall,nysmall;

  /*************************/
  /* parse input arguments */
  /*************************/
  for(k=1; k<argc; k++) {
    if (strcmp(argv[k],"-v") == 0) vflag = 1;
    else if (strcmp(argv[k],"-noshift") == 0) noshiftflag = 1;
    else if (strcmp(argv[k],"-ftype") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -ftype requires an argument\n\n"); return EXIT_FAILURE; }
      else { datatype = argv[k+1]; k++; }
    }
    else if (strcmp(argv[k],"-A") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -A requires an argument\n\n"); return EXIT_FAILURE; }
      else { fname_A = argv[k+1]; k++; }
    }
    else if (strcmp(argv[k],"-s") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -s requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	sigma_value = argv[k+1]; err = sscanf(sigma_value,"%lf",&sigma); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -s\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-e") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -e requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	eps_value = argv[k+1]; err = sscanf(eps_value,"%lf",&eps); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -e\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-p") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -p requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	peak_value = argv[k+1]; err = sscanf(peak_value,"%lf",&peakval); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -p\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-M") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -M requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	M_value = argv[k+1]; err = sscanf(M_value,"%d",&Nx); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -M\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-zx") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -zx requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	zx_value = argv[k+1]; err = sscanf(zx_value,"%lf",&zx); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -zx\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-N") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -N requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	N_value = argv[k+1]; err = sscanf(N_value,"%d",&Ny); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -N\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-zy") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -zy requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	zy_value = argv[k+1]; err = sscanf(zy_value,"%lf",&zy); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -zy\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (NULL == m_value) {
      m_value = argv[k]; err = sscanf(m_value,"%d",&nx); 
      if(err != 1) { display_usage(); printf("Error: could not retrieve properly input argument 'm'\n\n"); return EXIT_FAILURE; }
    }
    else if (NULL == n_value) {
      n_value = argv[k]; err = sscanf(n_value,"%d",&ny); 
      if(err != 1) { display_usage(); printf("Error: could not retrieve properly input argument 'n'\n\n"); return EXIT_FAILURE; }
    }
    else if (NULL == fname_T) fname_T = argv[k];
    else { display_usage(); printf("Error: could not retrieve properly input arguments.\n\n"); return EXIT_FAILURE; }
  }

  /********************************************/
  /* check consistency and set default values */
  /********************************************/
  if(NULL != M_value && NULL != zx_value) { display_usage(); printf("Error: options -M and -zx cannot be used simultaneously\n\n"); return EXIT_FAILURE; }
  if(NULL != N_value && NULL != zy_value) { display_usage(); printf("Error: options -N and -zy cannot be used simultaneously\n\n"); return EXIT_FAILURE; }
  if(NULL != zx_value && zx < 1.) { display_usage(); printf("Error: input 'zx' must be greater than or equal to one.\n\n"); return EXIT_FAILURE; }
  if(NULL != zy_value && zy < 1.) { display_usage(); printf("Error: input 'zy' must be greater than or equal to one.\n\n"); return EXIT_FAILURE; }
  if(NULL == fname_A && NULL == sigma_value) { display_usage(); printf("Error: at least one option must be used among -A and -s.\n\n"); return EXIT_FAILURE; }

  if ((NULL == m_value) || (NULL == n_value) || (NULL == fname_T)) {
    display_usage();
    if (NULL == m_value) printf("Error: input 'm' is missing\n\n");
    else if (NULL == n_value) printf("Error: input 'n' is missing\n\n");
    else printf("Error: input 'T' is missing\n\n");
    return EXIT_FAILURE;
  }

  if((NULL == datatype)||(strcmp(datatype,"float")==0)) type = 0;
  else if (strcmp(datatype,"uint8")==0) type = 8;
  else if (strcmp(datatype,"int8")==0) type = -8;
  else if (strcmp(datatype,"uint16")==0) type = 16;
  else if (strcmp(datatype,"int16")==0) type = -16;
  else if (strcmp(datatype,"uint32")==0) type = 32;
  else if (strcmp(datatype,"int32")==0) type = -32;
  else { display_usage(); printf("Error: unrecognized value for option -ftype.\n\n"); return EXIT_FAILURE; }

  if(NULL == fname_A && datatype) { printf("WARNING: option -ftype is useless without option -A\n"); };
  if(NULL == fname_A && noshiftflag) { printf("WARNING: option -noshift is useless without option -A\n"); };  
  if(NULL == sigma_value && peak_value) { printf("WARNING: option -p is useless without option -s\n"); };
  if(NULL == peak_value) peakval = 255.; 
  if(NULL == eps_value) eps = 1e-9; 

  /******************************************************************/
  /* retrieve dimensions of the high-resolution image domain (M, N) */
  /******************************************************************/
  if(M_value && Nx < nx) {
    display_usage();
    printf("Error: input 'M' must be greater than or equal to 'm' (M >= %d).\n\n",nx);
    return EXIT_FAILURE;
  }
  if(N_value && Ny < ny) {
    display_usage();
    printf("Error: input 'N' must be greater than or equal to 'n' (N >= %d).\n\n",ny);
    return EXIT_FAILURE;
  }

  if(NULL == zx_value && NULL == M_value) { zx = 2.; Nx = 2*nx; }
  else if(NULL == zx_value) zx = (double)Nx/(double)nx;
  else {
    Nx = (int)round(nx*zx);
    zx_tmp = zx;
    zx = (double)Nx/(double)nx;
    if(zx_tmp != zx) printf("WARNING: zx has been changed into %.17g (instead of zx=%g) to allow integer width for the high-resolution image domain\n",zx,zx_tmp);
  }

  if(NULL == zy_value && NULL == N_value) { zy = 2.; Ny = 2*ny; }
  else if(NULL == zy_value) zy = (double)Ny/(double)ny;
  else {
    Ny = (int)round(ny*zy);
    zy_tmp = zy;
    zy = (double)Ny/(double)ny;
    if(zy_tmp != zy) printf("WARNING: zy has been changed into %.17g (instead of zy=%g) to allow integer height for the high-resolution image domain\n",zy,zy_tmp);
  }

  if(vflag) {
    printf("\nRetrieve image domain dimensions\n");
    printf("================================\n");
    printf("width of the low-resolution domain: m = %d\n",nx);
    printf("height of the low-resolution domain: n = %d\n",ny);
    printf("width of the high-resolution domain: M = %d\n",Nx);
    printf("height of the high-resolution domain: N = %d\n",Ny);
    printf("horizontal super-resolution factor: zx = M/m = %g\n",zx);
    printf("vertical super-resolution factor: zy = N/n = %g\n",zy);
    printf("\n");
  }

  /**************************************/
  /* load the sequence of displacements */
  /**************************************/
  getasciitranslations(&dx,&dy,&ntranslations,fname_T,vflag);

  /*********************/
  /* memory allocation */
  /*********************/
  if(NULL != fname_A) { ASSERT_ALLOC(A = (double*) calloc (Nx*Ny,sizeof(double))); }
  
  /**********************/
  /* CORE OF THE MODULE */
  /**********************/

  /* compute gx = greatest common divisor of Nx and nx */
  u = Nx; v = nx; 
  while (v != 0) {
    r = u % v;
    u = v;
    v = r;
  }
  gx = u;
  
  /* compute gy = greatest common divisor of Ny and ny */
  u = Ny; v = ny; 
  while (v != 0) {
    r = u % v;
    u = v;
    v = r;
  }
  gy = u;

  /* compute domains HR & LR domain withsmallest dimensions (Nxsmall,Nysmall) & (nxsmall,nysmall) such as Nxsmall/nxsmall = Nx/nx & Nysmall/nysmall = Ny/nu */
  nxsmall = nx/gx;
  nysmall = ny/gy;
  Nxsmall = Nx/gx;
  Nysmall = Ny/gy; 

  /* memory allocation */
  ASSERT_ALLOC(Asmall = (double*) calloc (Nxsmall*Nysmall,sizeof(double)));
  
  /* perform error prediction (shift (0,0) frequency at the center of Asmall) */
  error_prediction(Asmall,&psnr,&mse,dx,dy,(sigma_value)?&sigma:NULL,nxsmall,nysmall,Nxsmall,Nysmall,ntranslations,peakval,eps,vflag,0);

  /* (optional) printf predicted PSNR and MSE */
  if(sigma_value) {
    if(vflag) printf("Predicted reconstruction quality\n================================\n");
    printf("predicted MSE = %.4e\n",mse);   
    printf("predicted PSNR (dB) = %02.4g\n",psnr);   
  }

  /* (optional) zoom Asmall by factor (gx,gy) using the nearest neighbor interpolation (+ unshift frequency domain if needed) */
  if(A) {
    if(noshiftflag) { // zoom + unshift 
      for(y=0;y<Ny;y++)
	for(x=0;x<Nx;x++)
	  A[((Ny+(y-Ny/2))%Ny)*Nx + (Nx+(x-Nx/2))%Nx] = Asmall[(y/gy)*Nxsmall+(x/gx)]; 
    }
    else { // zoom only
      for(adr=y=0;y<Ny;y++)
	for(x=0;x<Nx;x++,adr++)
	  A[adr] = Asmall[(y/gy)*Nxsmall+(x/gx)]; 
    }
  }

  /*********************************************************/
  /* if needed, save output (high-resolution) TIFF image A */
  /*********************************************************/
  if(fname_A && EXIT_FAILURE == save_monopage_striped_tiff_from_double(fname_A,A,Nx,Ny,0,type,"w")) {
    free(A); free(Asmall); free(dx); free(dy);
    return EXIT_FAILURE;
  }
  
  /***************/
  /* free memory */
  /***************/
  free(Asmall);  
  free(dx);
  free(dy);
  if(fname_A) free(A); 

  return EXIT_SUCCESS;
}
